torchtext/prototype/t5/modules.py

```diff
@@ -0,0 +1,819 @@
+            # residual connection
+            x = x + attn_out
+            x = x + self._ff_block(self.norm2(x))
+        else:
+            attn_out, position_bias = self._sa_block(x, src_mask, src_key_padding_mask, position_bias)
+            x, position_bias = self.norm1(x + attn_out)
+            x = self.norm2(x + self._ff_block(x))
+
+        return x, position_bias
+
+    # self-attention block
+    def _sa_block(
+        self,
+        x: Tensor,
+        attn_mask: Optional[Tensor],
+        key_padding_mask: Optional[Tensor],
+        position_bias: Optional[Tensor],
+    ) -> Tensor:
+        attn = self.self_attn(
+            x,
+            x,
+            x,
+            attn_mask=attn_mask,
+            key_padding_mask=key_padding_mask,
+            need_weights=False,
+            compute_relative_attention_bias=self.compute_relative_attention_bias,
+            relative_attention_num_buckets=self.relative_attention_num_buckets,
+            relative_attention_max_distance=self.relative_attention_max_distance,
+            relative_attention_bias=self.relative_attention_bias,
+            position_bias=position_bias,
+        )
+
+        x = attn[0]
+        if self.compute_relative_attention_bias and position_bias is None:
+            position_bias = attn[2]
+
+        return self.dropout1(x), position_bias
+
+    # feed forward block
+    def _ff_block(self, x: Tensor) -> Tensor:
+        x = self.linear2(self.dropout(self.activation(self.linear1(x))))
+        return self.dropout2(x)
+
+
+class T5Encoder(nn.TransformerEncoder):
+
+    r"""TransformerEncoder is a stack of N encoder layers. Users can build the
+    BERT(https://arxiv.org/abs/1810.04805) model with corresponding parameters.
+    Args:
+        encoder_layer: an instance of the TransformerEncoderLayer() class (required).
+        num_layers: the number of sub-encoder-layers in the encoder (required).
+        norm: the layer normalization component (optional).
+        enable_nested_tensor: if True, input will automatically convert to nested tensor
+            (and convert back on output). This will improve the overall performance of
+            TransformerEncoder when padding rate is high. Default: ``True`` (enabled).
+    Examples::
+        >>> encoder_layer = T5EncoderLayer(d_model=768, nhead=12, dim_feedfoward=3072, dropout=0.1, activation='relu', batch_first=True)
+        >>> t5_norm = T5LayerNorm(d_model=768)
+        >>> t5_encoder = T5Encoder(encoder_layer, num_layers=12, norm=t5_norm)
+        >>> src = torch.rand(10, 32, 512)
+        >>> out = t5_encoder(src)
+    """
+
+    def __init__(
+        self,
+        encoder_layer,
+        relative_attention_num_buckets,
+        num_heads,
+        num_layers=12,
+        norm=None,
+        enable_nested_tensor=True,
+    ):
+        super(T5Encoder, self).__init__(encoder_layer, num_layers, norm, enable_nested_tensor)
+
+        first_layer = copy.deepcopy(encoder_layer)
```

Comment:
I am not sure if the use of deep copy is guaranteed to keep working with PyTorch modules.

Instead of expecting an instance, it is more elegant to accept parameters required to build the component and instantiate encoders as much as needed.